오픈마켓 II [Step 1] Gundy, Wonbi

wonbi92 opened the pull request:
@wonhee009 
안녕하세요 라자냐! 
PR이 너무 늦어졌습니다.. ㅠㅠ 죄송합니다 ㅠㅠ
오픈마켓의 악명은 익히 들어 알고 있었지만 실제로 접해보니 악독하더구만요...

## ✓ 진행단계 : II Step-1

## 📝 STEP 진행 중 경험하고 배운 것
- 모던 컬렉션 뷰 사용하기
    ☑️ 커스터마이징 셀 구현하기
    ☑️ 가로스크롤 컬렉션 뷰 구현하기
- UIImagePickerController 사용하기
    ☑️ 사진첩에 접근하기
- UITextField, UITextView를 통해 사용자 입력받기
    ☑️ placeholder 사용하기


## 💭 고민한 부분

### 키보드 사용시 화면 스크롤링
- 이번 프로젝트의 가이드라인에는 키보드가 작성중인 컨텐츠를 가리면 안된다는 조건이 있었습니다. 기기별로 저마다의 컨텐츠 위치가 다를텐데 이를 적절히 스크롤해주는 것이 쉽지 않았습니다.
- 저희는 텍스트 필드를 입력중일 때는 imageCollectionView의 아랫부분으로, 텍스트 뷰를 입력중일 때는 productStackView의 아랫부분으로 오프셋을 정하여 적절한 위치로 화면이 이동하게 하였습니다.

### 코드 재사용
- 얼럿을 띄우는 메서드인 showAlert의 경우 action을 따로 분리하지 않고 코드를 재사용할 수 있도록 매개변수 message의 내용에 따라 `dismiss(animated: true)`할 수 있도록 조건문을 사용하였습니다.

### Entering data
- Apple HIG에 따르면 여러 텍스트 필드 등을 입력해야 할 때 동적으로 값을 검증하고 미리 안내하는 것이 사용자에게 좋다고 안내합니다. 저희도 이를 지키기 위해 필수 입력값인 이름, 가격, 설명 등에 대해서 입력하지 않은 경우 빨간색 보더라인을 추가하는 것으로 시각적 효과를 주고 있습니다.

> **Dynamically validate field values.**
> People can get frustrated when they have to go back and correct mistakes after filling out a lengthy form. When you verify values as soon as people enter them — and provide feedback as soon as you detect a problem — you give them the opportunity to correct errors right away. For numeric data in particular, consider using a number formatter, which automatically configures a text field to accept only numeric values. You can also configure a formatter to display the value in a specific way, such as with a certain number of decimal places, as a percentage, or as currency.

### 이미지 추가와 삭제
- 이미지를 처음에는 추가하지만 추가된 이미지를 다시 클릭하면 삭제되도록 로직을 구성해, 사용자가 이미지를 잘못 선택했을 때 cnacel을 눌러 화면에서 벗어났다가 다시 들어오는 불편함이 없도록 하였습니다.

## ❓ 조언을 얻고 싶은 부분

### 뷰 컨트롤러가 너무 방대함
- 코드를 작성하다 보니 뷰 컨트롤러가 400줄이 넘어가는 사태가 발생하였습니다.. MVC패턴의 단점이 뷰 컨트롤러가 너무 방대해진다는 것은 알고 있었지만, 그럼에도 이 패턴을 지키면서 뷰 컨트롤러를 가볍게 할 수 있는 방법이 있을까요? 저희가 생각 하기에는 뷰의 컴포넌트들을 따로 분리하여 구현하고 그 컴포넌트들을 구현한 뷰를 뷰컨의 뷰로 바꿔치기 하는 등의 방법을 생각했었습니다. 
```swift
class MainView: UIView {
    // 여기에 컴포넌트들이 들어가고 구현됩니다.
    // Label, stack, collection 등등..
}

class ViewController: UIViewController {
    override func viewDidLoad() {
        // 여기서 뷰가 교체됩니다.
        view = MainView()
    }
}
```
- 위와 같은 방법으로 뷰를 그리는 부분을 분리하는 방법을 생각했었습니다. 
- 이번 저희의 PR이 너무 늦어져서, 일단 먼저 PR을 보낸 후 라자냐의 조언을 얻고 step2에서 이 방식으로 리팩토링 해볼 계획입니다. 
- 이러한 방법이 괜찮은 방법일까요? 또, 방대해진 뷰 컨트롤러의 부담을 덜어주기 위한 다른 방법이나 트릭들이 있을지 궁금합니다!

### 텍스트 필드와 키보드 타입
- 숫자를 입력받는 곳에는 소수를 입력받는지 정수를 입력받는지에 따라 decimalPad와 numberPad로 구분하여 키보드를 제공하고 있습니다.
- 하지만 블루투스 키보드 등을 연결하고 사용한다면 해당 텍스트 필드에서도 숫자 이외의 값을 입력할 수 있을 것입니다.
- 이런 경우 사용자의 입력을 강제할 수 있도록 다른 조치를 취해야 할까요?

### 화면의 다른 부분 터치시 키보드 내리기
- 텍스트 필드나 텍스트 뷰를 edit하다가 키보드를 내릴 때 화면을 터치하는 것이 일반적으로 사용되는 UX라고 생각합니다.
- view의 `touchesBegan`을 통해 `view.endEditing(true)`를 호출하여 키보드를 내리려고 했는데, imageCollectionView 부분을 터치하는 경우 키보드가 내려가지 않았습니다.
- 해당 부분을 터치하더라도 키보드가 내려갈 수 있도록 탭 제스처를 추가했더니 셀을 선택할 수 없는 문제가 발생하였습니다.
- hitTest를 수정하는 방향으로도 접근하여서 첫 번째 셀을 터치하면 셀이 터치가 되고, 컬렉션뷰의 배경을 터치하면 키보드가 내려가게끔 로직을 구현하였습니다. 하지만 이미지가 추가됨에 따라 셀이 늘어나자 해당 로직대로는 셀을 터치할 수가 없어서 결국 터치로 키보드를 내리지 않고 Done 버튼을 추가하는 것으로 문제를 해결하였습니다.
- 이런 상황에서 터치로 키보드를 내리는 방법에 대한 조언을 얻고 싶습니다! 아래는 당시 구현한 코드...인것같습니다.
```swift
extension UICollectionView {
    open override func hitTest(_ point: CGPoint, with event: UIEvent?) -> UIView? {
        var touchedCell = superview?.hitTest(point, with: event)
        visibleCells.forEach { cell in
            if let cell = cell.hitTest(point, with: event) {
                 touchedCell = cell
            }
        }
        return touchedCell
    }
}
```

### 이미지를 용량 기준으로 리사이징하기
- 이번 프로젝트 조건 중, 이미지의 용량이 300KB가 넘으면 이를 리사이징하여 300이하가 되도록 구현하는 부분이 있었습니다.
- 저희는 첫번째로, 이미지를 jpeg로 압축하여 이미지의 용량을 조절하는 방법을 구현해 보았었습니다. 
    - 이 방법은, 이미지를 jpegData(compressionQuality:)메서드로 압축하되, 압축 퀄리티를 점점 낮춰서 용량이 300이하가 될 때까지 재귀하는 방법이었습니다.
```swift
func compress() -> Data? {
    var compressingValue: Double = 1.0
    
    while /*용량이 300이하일 때 까지*/ {
        if let data = self.jpegData(compressionQuality: compressingValue) {
            if /*data의 크기가 300 보다 크면*/ {
                compressingValue -= 0.1
            } else {
                image = data
            }
        }
    }        
}
```
- 간단히 작성해본 그때의 코드입니다. 이 방법의 단점은, 이미지를 계속해서 데이터로 인코딩을 진행해야 해서 앱의 속도가 느려지고, 이미지의 용량이 너무 클 경우 `compressionQuality` 매개변수 값이 음수가 될 때까지 반복된다는 점이었습니다.
- 이런 문제를 해결하기 위해, 이미지의 용량을 체크해서 300보다 클 경우, 300보다 얼만큼 큰지를 비율로 계산해 그 비율만큼 사진의 크기를 리사이징 하고 압축하는 방법을 생각했습니다.
    - 이 방법은, 이미지의 용량를 비율로 계산하여 클수록 이미지를 더 작게 만들어서 한번만 압축하는 방법이었습니다.
```swift
func updateImage(image: UIImage?) {
    guard let dataSize = image?.jpegData(compressionQuality: 0.9)?.count else { return }
    let kbSize = Double(dataSize / 1024)
    plusImageView.isHidden = true
    
    if kbSize > 300 {
    // 여기서 300보다 클 경우 scale을 계산하여 리사이징을 진행합니다.
        let scale = floor(sqrt(300 / kbSize) * 10) / 10
        productImageView.image = image?.resize(scale)
    } else {
        productImageView.image = image
    }
}

func resize(_ scale: Double) -> UIImage? {
    let target = CGRect(x: 0, y: 0, width: (self.size.width * scale), height: (self.size.height * scale))
    // 이미지를 target을 기준으로 리사이징 하는 로직이 들어갑니다.
    return newImage
    }
}
```
- 이 방법은 한번만 압축을 진행하기에 속도가 빨라 앱에 부담을 덜어주는 방법이었습니다. 하지만 이미지의 용량이 300보다 더 작아지는 경우도 생겼습니다.
- 결과적으로는 두번째 방법을 채택하여 이미지를 리사이징 하게 되었습니다. 다만, 이는 이미지의 용량이 300에 딱 떨어지게 해서 업로드 가능한 최대 화질의 이미지가 아니기 때문에 완벽하게 요구사항을 만족하지 못했다 판단하였습니다.
- 저희가 생각하지 못한 이미지를 용량을 기준으로 리사이징 하는 다른 방법이 또 있을까요? 그리고, 현업에서는 이러한 작업을 보통 어떻게 진행하는지 궁금합니다!

Gundy93 committed:
feat :POST를 위한 NewProduct 모델 생성 및 상품 등록/수정 화면으로의 전환 구현

- URLSessionProtocol에 URLRequest를 매개변수로 받는 dataTask(with:completionHandler:) 메서드 구현
- NetworkRequest에 httpHeader와 httpBody 추가
- Encodable을 채택하는 NewProduct 구조체 추가
- 상품 등록의 request를 담당하는 GenerateProductRequest 추가

wonbi92 committed:
feat : GenerateProductRequest에 httpHeader, httpBody를 만드는 로직 추가

- NewProduct의 CodingKeys 추가 및 정리
- NetworkError파일 분리
- Data, JSONEncoder 익스텐션 생성
- NetworkManager의 fetchData메서드 로직 수정
- NetworkManager에 postData메서드 추가

Gundy93 committed:
feat : collectionView의 refreshControl 추가

- refreshControl의 addTarget을 위한 메서드 refresh() 구현

wonbi92 committed:
feat : patch를 위한 Request객체 생성

- EditProduct모델 생성
- EditProductRequest객체 생성
- Request객체들이 상수만을 가지도록 리팩토링

Gundy93 committed:
feat : 상품 삭제를 위한 URISearchRequest 및 ProductDeleteRequest 구현

- EditProduct 수정
- 일부 리퀘스트 이름 변경

wonbi92 committed:
feat : AddProductViewController에 imageCollectionView추가

- 모던컬렉션뷰 방식을 사용하여 컬렉션 뷰 추가
- 이미지 업로드를 하는 ImageCell 커스텀셀 구현
- DEPLOYMENT_TARGET 을 14.0으로 변경

Gundy93 committed:
refactor : 접근제어자 수정

wonbi92 committed:
fix : 네트워킹이 실패했을 때 스크롤 리프레싱이 진행되지 않던 문제 수정

- 컬렉션 뷰의 Cell 네이밍 변경
- 당겨서 새로고침 진행중 스크롤을 밑으로 내렸을 때 앱이 튕기는 문제 수정
- 폴더 계층구조 정리
- 오타 수정

wonbi92 committed:
fix : Info위치로 인한 빌드 실패 오류 수정

Gundy93 committed:
docs : README.md 업데이트

wonbi92 committed:
docs : 리드미 수정

Gundy93 committed:
design : 상품 등록/수정 화면 UI 구성

wonbi92 committed:
feat : UIImagePickerController를 통한 이미지 피커 구현

- 인덴트 및 개행 정리
- ImageCell에 셀을 리셋하는 메서드 추가

Gundy93 committed:
feat : 각 TextField에 적절한 keyboard 호출 구현

- imageCellIdentifiers 오타 수정

wonbi92 committed:
feat : 입력값이 API에서 요구하는 값인지 체크하는 로직 구현

- checkCanAddProduct메서드 구현
- 입력값이 요구하는 값이 아닐 경우 테두리가 빨간색이 되도록 구현

Gundy93 committed:
feat : 이미지 용량 제한에 따른 리사이징 기능 추가

- UIImage extension 추가

wonbi92 committed:
feat : 키보드가 올라올 때 컨텐츠를 가리지 않도록 로직 구현

- 스크롤이 가능하도록 뷰 전체의 컴포넌트를 담는 스크롤 뷰 추가
- textFieldDidBeginEditing, textViewDidBeginEditing에서 키보드가 컨텐츠를 가리지 않게 스크롤하는 로직 추가
- 키보드 입력을 완료했을 때 키보드를 내릴 수 있는 Done버튼 추가

Gundy93 committed:
feat : 상품 등록 페이지에서 Done 버튼을 누르면 상품이 등록되는 로직 구현

- networkManager 프로퍼티 생성
- 등록을 시작하는 postProductData 메서드 생성
- 필수 입력이 아닌 discountedPrice, stock의 텍스트 필드가 비었는지 확인하고 0으로 변경하는 checkTextField 메서드 추가
- JSON 데이터를 만드는 makeNewProductData 메서드 추가
- 이미지를 한 Data에 모으는 makeImageData 메서드 추가
- ImageCell에 자신의 이미지 데이터를 반환하는 productImageData 메서드 추가

wonbi92 committed:
refactor : 이미지 리사이징이 적용안되는 문제 수정

- 더이상 UIImage의 resize메서드가 @discardableResult를 채택하지 않음

Gundy93 committed:
refactor : 코드 컨벤션 정리 및 접근제어자 수정

wonhee009 commented on OpenMarket/OpenMarket/Model/EditProduct.swift:
```diff
@@ -0,0 +1,47 @@
+
+import Foundation
+
+struct EditProduct: Codable {
```
EditProduct, NewProduct, Product 이렇게 3개의 struct가 있는데요.
각각 어떤 역할을 할때 사용하는걸까요?

wonbi92 replied:
Product는 상품 리스트 페이지에서 개별 상품의 정보를 주기 위한 모델,
EditProduct는 상품 수정페이지에서 뷰에 상품 정보를 주고, 변경점을 서버에 전달하기 위한 모델,
NewProduct는 상품 추가 페이지에서 사용하는 모델입니다!

사실, EditProduct의 경우 아직 구현만 되어있고 사용을 하진 않지만 이후의 스텝을 위해 미리 모델링을 해두었습니다!

wonhee009 replied:
비슷한 프로퍼티들이 반복되고 있는거 같은데요. 따로 정의한 이유가 있을까요?

Gundy93 replied:
비슷하지만 리퀘스트마다 요구하는 조건이나 응답으로 전달되는 데이터의 프로퍼티가 조금씩 다르기 때문에 서로 다르다고 생각했습니다. 혹시 이러한 유사성을 보이는 경우 다른 적절한 방법이 있을까요?

wonhee009 replied:
EditProduct, NewProduct의 경우에는 동일한 타입 사용할 수 있을거라고 생각했습니다.
수정/추가 상황에서 요구하는 프로퍼티가 동일하거나 한두개 정도 다를거 같고 (id 같은 경우가 edit은 필수일거 같고, new는 필요 없는 경우일거 같은데요. 이런 경우는 optional로 설정) 사실상 같은 Product에 대한거니 동일한 타입을 사용해야 한다고 생각했습니다.
Product의 경우 화면을 그리기 위한 Model로 이해했는데요. 사실상 얘도 수정/추가에서 사용하는 Model과 동일한 객체를 가르킨다고 생각했습니다.
```swift
struct ProductResponse {
   let id: String?
   let name: String
   ...
}

// 상품 리스트 페이지
class ProductListVC {
   struct Product {
      화면 그리는데 필요한 프로퍼티
   }
   ...
}
```
이런 식으로 필요한 부분만 뽑아다 쓸 수 있을거 같습니다.
이 부분은 이후 스텝에서 고민해주세요.

wonhee009 commented on OpenMarket/OpenMarket/Network/OpenMarketNetworkRequest.swift:
```diff
@@ -5,37 +5,140 @@
 //  Created by Gundy, Wonbi on 2022/11/15.
 //
 
+import UIKit
```
UIKit을 import해야하는 이유가 있나요?

wonbi92 replied:
원래, UIImage를 이니셜라이저의 매개변수로 받아서 데이터 작업을 하도록 로직을 구성했었는데, 리팩토링 과정에서 그 부분이 빠졌습니다. 미처 발견하지 못하고 Foundation으로 바꾸지 못했었네요.. Foundation으로 바꿔두었습니다!

wonhee009 commented on OpenMarket/Network/networkError.swift:
```diff
@@ -0,0 +1,31 @@
+        case .invalidURL:
+            return "invalidURL"
+        }
+    }
```
[LocalizedError](https://developer.apple.com/documentation/foundation/localizederror)에 대해 알고 계실까요?

wonbi92 replied:
이전 스텝의 `DecodingError`에서 말씀해주신 부분처럼 `NetworkError`가 `LocalizedError`를 채택하고 `errorDescription`을 구현하는 쪽으로 리팩토링을 진행해 보았습니다! 사용할 때도 `.localizedDescription`을 사용하도록 리팩토링 하였습니다!

wonhee009 commented on OpenMarket/OpenMarket/GridCell.swift:
```diff
@@ -105,7 +105,8 @@ final class GridCollectionViewCell: UICollectionViewCell {
     }
     
-    func configureCell(from product: Product) {
+    func configureCell(from product: Product?) {
```
product를 옵셔널로 받고 있는데요.
`configureCell` 호출부에서 옵셔널 바인딩을 한 후 `configureCell(from product: Product)`을 호출하는건 어떤까요?

wonbi92 replied:
아하! 이부분은 저희가 첨자 문법에서 `[valid: ]`를 신경쓰다보니 약간 요상한 로직으로 코드가 짜여진 것 같습니다.. 😅😅 
`configureCell`메서드는 넌 옵셔널 Product를 받고 호출부에서 이를 바인딩 하는 방식으로 코드를 리팩토링 하였습니다!

wonhee009 commented on OpenMarket/OpenMarket/GridCell.swift:
```diff
@@ -105,7 +105,8 @@ final class GridCollectionViewCell: UICollectionViewCell {
     }
     
-    func configureCell(from product: Product) {
+    func configureCell(from product: Product?) {
```
해당 메서드 호출할때에는 `cell.configureCell(_)`로 호출될텐데요. `configure(from product)`여도 충분히 의미 전달이 될거 같습니다.

wonbi92 replied:
넵! cell이 겹치는 부분이 있네요! 수정하였습니다!

wonhee009 commented on OpenMarket/OpenMarket/AddProductViewController.swift:
```diff
@@ -8,8 +8,488 @@
 
 final class AddProductViewController: UIViewController {
+    private var imageCellIdentifiers: [Int] = [0]
+    private let defaultIdentifier: Set<Int> = [0, 1, 2, 3, 4]
```
이 프로퍼티는 뭔가요?

wonbi92 replied:
이미지를 컬렉션 뷰 셀에 올리거나 셀을 삭제할 때 사용하는 `ItemIdentifier`로 사용하기 위해 선언해둔 프로퍼티 입니다. 

`defaultIdentifier`는 셀의 Identifier가 중복되어 선언되지 않도록 `Set<Int>` 로 선언하여 스냅샷의 `appendItems`메서드를 실행할 때 매개변수의 값을 `subtracting`메서드를 사용한 결과 값중 하나를 사용하도록 로직을 구성하였습니다.

`imageCellIdentifiers`는 이미 추가된 셀의 Identifier를 배열로 가지고 있어 위 `subtracting`메서드를 사용할 때 이미 append된 항목의 Identifier가 다시 append되지 않도록 해주는 역할입니다.

wonhee009 commented on OpenMarket/OpenMarket/AddProductViewController.swift:
```diff
@@ -8,8 +8,488 @@
+        return scrollView
+    }()
+    
+    private lazy var backView: UIView = {
```
lazy로 선언된 이유가 있나요?

wonbi92 replied:
원래, `UIStackView`로 진행하다가 `UIView`로 타입을 변경하게 되었습니다. 그 과정에서 실수로 lazy로 선언했던 부분이 남게 되었습니다.. let으로 리팩토링 하였습니다!

wonhee009 commented on OpenMarket/OpenMarket/AddProductViewController.swift:
```diff
@@ -8,8 +8,488 @@
+    }
+    
+    @objc private func tapDoneButton() {
+        if imageCellIdentifiers.count == 1 {
+            showAlert(message: "이미지를 추가해 주세요.")
+            return
+        }
+        
+        if !checkCanAddProduct() {
+            showAlert(message: "입력값이 잘못되었습니다.")
+            return
+        }
```
early exit 상황으로 보여 guard문 이용할 수 있을거 같습니다.

wonbi92 replied:
넵! early exit 에는 guard문을 통해 빠르게 탈출하도록 하는것이 좋습니다! guard문으로 리팩토링 하였습니다!!

wonhee009 commented on OpenMarket/OpenMarket/AddProductViewController.swift:
```diff
@@ -8,8 +8,488 @@
+        postProductData()
+    }
+    
+    private func postProductData() {
+        guard let params = makeNewProductData(),
+              let images = makeImageData(),
+              let request = ProductAddRequest(identifier: "c598a7e9-6941-11ed-a917-8dbc932b3fe4", params: params, images: images).request
+        else { return }
+        
+        networkManager.postData(form: request) { result in
+            switch result {
+            case .success(_):
+                DispatchQueue.main.async { [weak self] in
+                    self?.showAlert(message: "등록 성공!")
+                }
+            case .failure(let error):
+                print(error.localizedDescription)
+            }
+        }
+    }
```
`postProductData()`에서 `makeNewProductData, makeImageData`를 호출하면서 post에 필요한 데이터들을 만들어주고 있는데요.
`postProductData()`에서는 `request`만 받도록 데이터를 만드는 역할을 나누는건 어떤가요?

Gundy93 replied:
`postProductData()`에서 `request`만 받을 수 있도록 `makeProductAddRequest()` 메서드를 만들어서 `URLRequest`를 반환하도록 하였습니다.

wonhee009 commented on OpenMarket/OpenMarket/AddProductViewController.swift:
```diff
@@ -8,8 +8,488 @@
+        guard let discountedPrice = discountedPriceTextField.text,
+              let stock = stockTextField.text else { return }
+        
+        if discountedPrice.isEmpty {
+            discountedPriceTextField.text = "0"
+        }
+        if stock.isEmpty {
+            stockTextField.text = "0"
+        }
```
makeNewProductData에서 해당 메서드가 호출될거 같은데요.
discountedPrice과 stock이 empty라면 임의로 0으로 만들어 주는건가요?
이런 상황은 상품을 등록할 수 없는 상황이 아닌걸까요?

Gundy93 replied:
<img width="824" alt="스크린샷 2022-12-07 오전 11 14 41" src="https://user-images.githubusercontent.com/106914201/206071150-e4d0cb83-a147-4240-a3ba-c206411e4225.png">

API에 따르면 두 값은 default가 0이기 때문에 사용자가 지나치더라도 기본값인 0을 넣을 수 있도록 하였습니다!

wonhee009 replied:
해당 메서드는 textField의 텍스트를 0으로 지정해주는거 같은데요.
첨부해주신 이미지는 API를 쏠때 해당 값들이 없다면 0으로 넣어달라는 얘기인거 같습니다.
textField는 그대로 유지하고 parameter를 만드는 부분에서 0으로 넣어줘야하지 않을까요?

Gundy93 replied:
시각적으로 0이 들어가는 편이 사용자 측면에서 이해하기 좋을거라고 생각해서 그렇게 했었습니다 🥲 텍스트필드는 건들지 않고 진행할 수 있도록 하겠습니다! 😃

wonhee009 commented on OpenMarket/OpenMarket/AddProductViewController.swift:
```diff
@@ -8,8 +8,488 @@
+    private func showAlert(message: String, completion: (() -> Void)? = nil) {
+        let alert = UIAlertController(title: nil, message: message, preferredStyle: .alert)
+        let action = UIAlertAction(title: "확인", style: .default) { _ in
+            if alert.message == "등록 성공!" { self.dismiss(animated: true) }
```
등록 완료후에는 해당 VC를 dismiss 하기 위한 코드로 이해했는데요.
dismiss 조건이 alert.message인 부분이 어색한거 같습니다.
등록이 완료되었다는 것을 Network result/response로 알 수 없을까요?

Gundy93 replied:
저희가 라자냐의 의도를 정확히 이해했는지 모르겠지만, result가 success인 경우 컴플리션 핸들러를 넣어주는 방법으로 `showAlert()` 메서드를 사용해 보았습니다.

wonhee009 replied:
음... 지금 showAlert을 여러 이미지 추가할때, 입력값이 잘못 되었을 때, 등록에 성공했을 때에서 호출해주고 있는거 같은데요.
이때에 등록에 성공했을 때에만 dismiss해주고 싶은거로 이해했습니다.
우선... completion을 추가했는데 completion을 부르는 부분이 어디 있을까요?
그리고 필요한 alert을 showAlert 하나로만 관리하는 부분은 좋으나 `if alert.message == "등록 성공!"`과 같은 코드는 지양했으면 좋겠습니다. `if alert.message == "등록 성공!" || "AA" || "BB"`와 같은 코드가 될 수도 있으니까요.

Gundy93 replied:
```swift
private func showAlert(message: String, completion: (() -> Void)? = nil) {
    let alert = UIAlertController(title: nil, message: message, preferredStyle: .alert)
    let action = UIAlertAction(title: "확인", style: .default) { _ in
        if let completion = completion {
            completion()
        }
    }
    
    alert.addAction(action)
    present(alert, animated: true)
}
```

showAlert 메서드의 매개변수로 받아서 내부에서 실행시키고 있습니다. 😃

```swift
case .success(_):
    DispatchQueue.main.async { [weak self] in
        self?.showAlert(message: "등록 성공!") {
            self?.dismiss(animated: true)
        }
    }
```
혹시 completion을 전달인자로 넣어주는 부분을 말씀하신거면 postProductData 메서드의 success 케이스에서  `dismiss(animated: true)`를 전달하고 있습니다.

wonhee009 commented on OpenMarket/OpenMarket/AddProductViewController.swift:
```diff
@@ -8,8 +8,488 @@
+}
+
+extension AddProductViewController {
+    private func createLayout() -> UICollectionViewLayout {
```
`createCollectionViewLayout`은 어떤가요? 현재 네이밍은 어떤 레이아웃을 만들고 있는지 명확하지 않은거 같습니다.

Gundy93 replied:
명확하지 않은 네이밍을 `createCollectionViewLayout`로 수정했습니다!

wonhee009 commented on OpenMarket/OpenMarket/AddProductViewController.swift:
```diff
@@ -8,8 +8,488 @@
+        return layout
+    }
+    
+    private func configureHierarchy() {
```
어떤 의미의 네이밍인가요?

Gundy93 replied:
collectionView의 delegate 및 layout 등을 설정하는 것을 계층 구조로 생각해 그렇게 적었는데, 모호한 네이밍인 것 같아 `configureCollectionView`로 수정하였습니다!

wonhee009 commented on OpenMarket/OpenMarket/AddProductViewController.swift:
```diff
@@ -8,8 +8,488 @@
+    private func checkCanAddProduct() -> Bool {
+        if let nameCount = nameTextField.text?.count,
+           !((3...100) ~= nameCount) {
+            nameTextField.layer.borderWidth = 1.0
+            nameTextField.layer.borderColor = UIColor.systemRed.cgColor
+            nameTextField.layer.cornerRadius = 5
```
아래에도 동일하게 반복되는 디자인 값들인거 같습니다.
상수로 들고 있는건 어떨까요?

Gundy93 replied:
`UIView`를 매개변수로 받는 `highlightTextBounds(_:)` 메서드를 정의해서 강조가 필요한 해당 사항에 각 텍스트 필드와 텍스트 뷰를 전달인자로 넣어주었습니다.

wonhee009 replied:
메서드가 어디에 있나요?

wonbi92 replied:
```swift
    private func highlightTextBounds(_ view: UIView) {
        let highlightedBorderWidth: CGFloat = 1.0
        let cornerRadius: CGFloat = 5
        
        view.layer.borderWidth = highlightedBorderWidth
        view.layer.borderColor = UIColor.systemRed.cgColor
        view.layer.cornerRadius = cornerRadius
    }
```
다음과 같이 메서드를 따로 만들어서 호출하는 방식으로 변경 해 보았습니다!

wonhee009 commented on OpenMarket/OpenMarket/ImageCell.swift:
```diff
@@ -0,0 +1,74 @@
+import UIKit
+
+final class ImageCell: UICollectionViewCell {
+    var isGetImage: Bool {
+        return productImageView.image != nil
+    }
```
이미지 유무에 관한 프로퍼티로 이해했는데요.
`canGetImage` 혹은 `imageIsNotEmpty` 같은 네이밍은 어떤가요?

Gundy93 replied:
`ImageCell`은 이미지만 가지고 있고, 타입의 이름에 이미지가 들어가기 때문에 `isNotEmpty` 정도로만 적어도 뜻이 전달될 것이라고 생각해 `isNotEmpty`로 네이밍 수정하였습니다.

wonhee009 commented on OpenMarket/OpenMarket/Network/OpenMarketNetworkRequest.swift:
```diff
@@ -5,37 +5,140 @@
+    init(identifier: String, params: Data, images: Data) {
+        var body = Data()
+        var header: [String: String] = [:]
+        let boundary = "----\(UUID().uuidString)"
+        let lineBreak = "\r\n"
+    
+        header["identifier"] = identifier
+        header["Content-Type"] = "multipart/form-data; boundary=\(boundary)"
+        
+        body.append("--" + boundary + lineBreak, using: .utf8)
+        body.append("Content-Disposition:form-data; name=\"params\"" + lineBreak, using: .utf8)
+        body.append("Content-Type: application/json" + lineBreak, using: .utf8)
+        body.append(lineBreak, using: .utf8)
+        body.append(params)
+        body.append(lineBreak, using: .utf8)
+        body.append("--" + boundary + lineBreak, using: .utf8)
+        body.append("Content-Disposition:form-data; name=\"images\"; filename=\"image.jpeg\"" + lineBreak, using: .utf8)
+        body.append("Content-Type: image/jpeg" + lineBreak, using: .utf8)
+        body.append(lineBreak, using: .utf8)
+        body.append(images)
+        body.append(lineBreak, using: .utf8)
+        body.append("--" + boundary + "--", using: .utf8)
```
해당 로직은 비즈니스 로직으로 느껴지는데 따로 분리하는건 어떤가요?
`NetworkRequest`에서는 해당 프로퍼티들을 세팅만 해줘야한다고 생각합니다.

Gundy93 replied:
리퀘스트를 생성하기 위한 header와 body를 갖는 MultipartFormData 추가하고, 이니셜라이저에 전달인자로 넣어줌으로써 `NetworkRequest`에서는 프로퍼티 세팅만이 이루어지게 수정하였습니다!

wonhee009 reviewed (changes requested):
수고하셨습니다:)

- VC가 커진 부분 확인해 보니 UI 요소들 때문으로 느껴졌는데요. MVC에서 UI로 인해 코드가 길어지는건 어쩔 수 없다고 생각합니다. MVVM을 사용하더라도 UI 코드가 많으면 VC가 방대해지는건 동일하거든요. 다만, MVVM에서는 비즈니스 로직이 VC에서 분리되었기 때문에 동일 코드에서 본다면 MVC보다는 가벼운거지 MVVM이라고 해당 문제가 없는건 아닙니다. View를 바꿔치면 코드는 줄일 수 있을거 같습니다.
- 키보드 관련해서는 기획서 스펙에 따라 달라지지 않을까 합니다.
- 저도 코드로 짜보진 않았는데 hitTest에서 키보드가 showing 중일 때에는 키보드 제외한 화면 탭시 키보드를 내려주면 되지 않을까요...?
- 이미지 리사이징 관련해서는 저도 안해봐서 한 번 확인해볼게요~

wonbi92 committed:
refactor : 코드 리팩토링 및 메서드 네이밍 수정

- Cell의 configureCell메서드의 네이밍 수정, 및 매개변수 타입 리팩토링
- NetworkError가 LocalizedError를 채택하도록 리팩토링

Gundy93 committed:
refactor : 로직 분리 및 네이밍 수정

- createLayout, configureHierarchy등을 createCollectionViewLayout, configureCollectionView와 같은 명확한 네이밍으로 수정
- 메서드가 하나의 기능을 갖도록 로직 분리
- 리퀘스트를 생성하기 위한 header와 body를 갖는 MultipartFormData 추가

wonhee009 commented on OpenMarket/Network/NetworkError.swift:
```diff
@@ -0,0 +1,31 @@
+    case invalidData
+    case invalidURL
+    
+    public var errorDescription: String? {
```
`public`여야 하나요?.?

wonbi92 replied:
LocalizedError 관련 예제를 찾아보던 중 Almofire가 이  LocalizedError를 채택하여 구현한 부분을 찾게 되어 그 부분을 참고해 코드를 작성하다 보니 저희도 모르게 `public`으로 선언하게 된 것 같습니다 😭 `public`접두사는 지웠습니다!

wonbi92 committed:
refactor : 할인가격이 상품가격보다 큰 경우 사용자에게 알리는 로직 추가

- AddProductView의 스크롤 뷰 오토레이아웃 충돌 수정
- 120자가 넘는 코드들 개행 정리
- ProductsViewController의 viewWillAppear재정의 삭제

Gundy93 committed:
refactor : errorDescription 접근제어자 수정, postProductData 로직 수정

- discountedPriceTextField와 stockTextField의 text를 변경하지 않게 로직 수정

wonhee009 reviewed (approved):
수고하셨습니다:)
